src/invite-api.js

```diff
@@ -1,236 +1,440 @@
+ * @internal
+ * @typedef {import('./generated/rpc.js').Invite} InviteInternal
+ */
+
+/**
+ * @internal
+ * @typedef {import('./generated/rpc.js').ProjectJoinDetails} ProjectJoinDetails
  */
 
 /**
- * @template K
- * @template V
- * @extends {Map<K, Set<V>>}
+ * @typedef {object} Invite
+ * @prop {string} inviteId
+ * @prop {string} projectName
+ * @prop {string} [roleName]
+ * @prop {string} [roleDescription]
+ * @prop {string} [invitorName]
  */
-class MapOfSets extends Map {
-  /** @param {K} key */
-  get(key) {
-    const existing = super.get(key)
-    if (existing) return existing
-    /** @type {Set<V>} */
-    const set = new Set()
-    super.set(key, set)
-    return set
+
+/**
+ * Manage pending invite state.
+ */
+class PendingInvites {
+  /**
+   * @internal
+   * @typedef {object} PendingInvite
+   * @prop {string} peerId
+   * @prop {InviteInternal} invite
+   * @prop {boolean} isAccepting
+   */
+
+  /** @type {Map<string, PendingInvite>} */
+  #byInviteId = new Map()
+
+  /** @type {Map<string, PendingInvite[]>} */
+  #byProjectPublicId = new Map()
+
+  /**
+   * @returns {Iterable<PendingInvite>} the pending invites, in insertion order
+   */
+  invites() {
+    return this.#byInviteId.values()
+  }
+
+  /**
+   * @param {PendingInvite} pendingInvite
+   * @throws if adding a duplicate invite ID
+   * @returns {void}
+   */
+  add(pendingInvite) {
+    const { invite } = pendingInvite
+    const { inviteId, projectPublicId } = invite
+
+    const inviteIdString = keyToId(inviteId)
+    assert(!this.#byInviteId.has(inviteIdString), 'Added duplicate invite')
+    this.#byInviteId.set(inviteIdString, pendingInvite)
+
+    const existingForProject = this.#byProjectPublicId.get(projectPublicId)
+    if (existingForProject) {
+      existingForProject.push(pendingInvite)
+    } else {
+      this.#byProjectPublicId.set(projectPublicId, [pendingInvite])
+    }
+  }
+
+  /**
+   * @param {Buffer} inviteId
+   * @returns {void}
+   */
+  markAccepting(inviteId) {
+    const inviteIdString = keyToId(inviteId)
+
+    const oldPendingInvite = this.#byInviteId.get(inviteIdString)
+    assert(!!oldPendingInvite, `Couldn't find invite for ${inviteIdString}`)
+
+    const newPendingInvite = { ...oldPendingInvite, isAccepting: true }
+
+    this.#byInviteId.set(inviteIdString, newPendingInvite)
+
+    const {
+      invite: { projectPublicId },
+    } = newPendingInvite
+    const newPendingInvitesForThisProject = (
+      this.#byProjectPublicId.get(projectPublicId) ?? []
+    ).map((pendingInvite) =>
+      pendingInvite.invite.inviteId.equals(inviteId)
+        ? newPendingInvite
+        : pendingInvite
+    )
+    this.#byProjectPublicId.set(
+      projectPublicId,
+      newPendingInvitesForThisProject
+    )
+  }
+
+  /**
+   * @param {Buffer} inviteId
+   * @returns {boolean}
+   */
+  hasInviteId(inviteId) {
```

Comment:
A helper could end up be handy here. Not sure if it's warranted/needed.

```js
const BufferMap extends Map {
  get(key) { return super.get(keyToId(key)) }
  set(key, value) { return super.set(keyToId(key), value) }
  has(key) { return super.get(keyToId(key)) }
  delete(key) { return super.delete(keyToId(key)) }
}
```

Reply:
Made a generic `HashMap` class in 8ad29aa2237cda365cc7b1e2653d94ca62e9ed6.